crates/bitcoin-da/tests/bitcoin_da_tests.rs

```diff
@@ -386,47 +439,149 @@ mod tests {
         }
     }
 
+
     #[test]
-    fn test_read_height() {
+    fn test_read_data_by_block_height() {
+        // ===============================
+        // Given: a block with embedded data
+        // ===============================
+    
+        // Prepare the data and relayer configuration
+        let embedded_data = b"Hello, height!";
         let relayer = Relayer::new(&Config::new(
-            // "YOUR_NODE_IP:38332".to_owned(), // SIGNET
-            "localhost:8332".to_owned(), // REGNET
+            NODE_IP.to_owned(),
             "rpcuser".to_owned(),
             "rpcpass".to_owned(),
         ))
         .unwrap();
-        let height = 112; // change to whatever height that contains a tx
-        match relayer.read_height(height) {
+    
+        // Obtain the current block height before embedding
+        let blockchain_info = relayer.client.get_blockchain_info().unwrap();
+        let current_height = blockchain_info.blocks;
+    
+        // Embed the data into the blockchain by writing a transaction
+        match relayer.write(embedded_data) {
+            Ok(txid) => {
+                println!("Txid: {}", txid);
+                println!("Successful write");
+            }
+            Err(e) => panic!("Write failed with error: {:?}", e),
+        }
+    
+        // Add the transaction to a new block
+        relayer.generate_blocks(1).unwrap();
+    
+        // Ensure there's enough time for the transaction to be processed
+        std::thread::sleep(std::time::Duration::from_secs(2));
+    
+        // ================================
+        // When: read data by block height
+        // ================================
+    
+        let read_result = relayer.read_height(current_height + 1);
+    
+        // ==========================================
+        // Then: assert outcomes and expected results
+        // ==========================================
+    
+        // Assert the data was correctly embedded and can be read back
+        match read_result {
             Ok(data) => {
-                // Change this line to whatever data you want.
-                // "bark" is appended to the beginning of the data
-                assert_eq!(data, b"barkHello, world!".to_vec());
+                assert!(data.windows(b"barkHello, height!".len()).any(|window| window == b"barkHello, height!"));
```

Comment:
not sure to understand why

Reply:
When I use:
`assert_eq!(data, b"barkHello, world!".to_vec());`

I get:
> thread 'tests::test_read_data_by_block_height' panicked at 'assertion failed: `(left == right)`
>   left: `[98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100, 33, 98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100, 33, 98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 104, 101, 105, 103, 104, 116, 33, 98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100, 33, 98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100, 33]`,
>  right: `[98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100, 33]`', crates/bitcoin-da/tests/bitcoin_da_tests.rs:490:17

It seems like there is more data than just "barkHello, world!". Not sure if I'm doing something incorrectly on my end.

Reply:
looks like it's the same data in loop, i'm not sure why

Reply:
Changed back to what you guys had in your `test_read_height` function:
        ```
        // Change this line to whatever data you want.
                // "bark" is appended to the beginning of the data
                assert!(
                    data == b"barkHello, world!".to_vec(),
                    "Assertion failed. This test is designed to be run manually. Expect it to fail in auto mode. Received data: {:?}",
                    data
                );
                println!("Successful read");
```

It works well when I run the test individually, but when I run all the tests seems like things are getting written to the block I'm trying to use from other tests. Will look into it

Reply:
I think I understand what is going on (Correct me if I'm wrong). The `read_height` function returns an appended vector of all transaction data from a single block. But the data we are sending could just be one of many transaction data, as opposed to the only piece of data in there. 

Assuming this is true, I decided to use a unique identifier in the embedded data to make sure we are retrieving only the data we sent.

As of now, changed back to this:
```
            Ok(data) => {
                // Check for the presence of the unique data in the block
                assert!(
                    data.windows(embedded_data.len()).any(|window| window == embedded_data.as_slice()),
                    "Unique data not found in block. Received data: {:?}",
                    data
                );
                println!("Successful read");
            }
            Err(e) => panic!("Read failed with error: {:?}", e),
```